pkg/spoke/addon/certificate_manager.go

```diff
@@ -0,0 +1,251 @@
+
+	// run a ClientCertificateController to create/rotate client certificate for addon
+	clientCertController := clientcert.NewClientCertificateController(
+		m.config.InstallationNamespace,
+		m.config.SecretName,
+		m.hubCSRInformer,
+		kubeInformerFactory.Core().V1().Secrets(),
+		m.kubeClient.CoreV1(),
+		m,
+		m.recorder,
+		fmt.Sprintf("AddonClientCertificateController@%s:%s", m.config.AddonName, m.config.SignerName),
+	)
+	go clientCertController.Run(ctx, 1)
+
+	if len(m.config.BootstrapSecretName) == 0 {
+		return nil
+	}
+
+	// if bootstrap secret specified, run a contoller to keep dumping secret into fs
+	secretDumpingController := clientcert.NewSecretDumpingController(
+		m.bootstrapKubeconfigDir,
+		m.config.InstallationNamespace,
+		m.config.BootstrapSecretName,
+		m.kubeClient.CoreV1(),
+		kubeInformerFactory.Core().V1().Secrets(),
+		m.recorder,
+		fmt.Sprintf("SecretDumpingController@%s:%s", m.config.AddonName, m.config.SignerName),
+	)
+	go secretDumpingController.Run(ctx, 1)
+
+	return nil
+}
+
+// stop stops controllers and informers it started and clean up resources it created.
+func (m *addonCertificateManager) stop() error {
+	if m.stopFunc != nil {
+		m.stopFunc()
+	}
+
+	// delete secret created
+	err := m.kubeClient.CoreV1().Secrets(m.config.InstallationNamespace).Delete(context.Background(), m.config.SecretName, metav1.DeleteOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// delete files dumped from bootstrap secret
+	return os.RemoveAll(m.bootstrapKubeconfigDir)
+}
+
+func (m *addonCertificateManager) Owner() string {
+	return fmt.Sprintf("addon (name=%s,signer=%s)", m.config.AddonName, m.config.SignerName)
+}
+
+func (m *addonCertificateManager) FilterCSREvents(obj interface{}) bool {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false
+	}
+	labels := accessor.GetLabels()
+	// only enqueue csr from a specific managed cluster
+	if labels[clientcert.ClusterNameLabel] != m.clusterName {
+		return false
+	}
+	// only enqueue csr created for a specific managed cluster
+	if labels[clientcert.AddonNameLabel] != m.config.AddonName {
+		return false
+	}
+
+	// only enqueue csr with a specific signer name
+	csr, ok := obj.(*certificates.CertificateSigningRequest)
+	if !ok {
+		return false
+	}
+	if *csr.Spec.SignerName != m.config.SignerName {
+		return false
+	}
+	return true
+}
+
+func (m *addonCertificateManager) GetCSRClient() (csrclient.CertificateSigningRequestInterface, error) {
+	// if addon has no bootstrap secret, always return the default csr client (created with registration
+	// agent's hub kubeconfig)
+	if len(m.config.BootstrapSecretName) == 0 {
+		return m.hubCSRClient, nil
+	}
+
+	// if the addon has already had a valid client cert (bootstraped), return the default
+	// csr client for cert renewal
+	secret, err := m.kubeClient.CoreV1().Secrets(m.config.InstallationNamespace).Get(context.Background(), m.config.SecretName, metav1.GetOptions{})
```

Comment:
I think we alway use the registration csr client? How can we ensure that the addon's client cert can be used for csr.